### PR TITLE
[libcollections] Fix unused import warnings during stage0 build

### PR DESCRIPTION
**Pull request by wesleywiser**
Fixes a bunch of warnings that are generated during the stage0 build of libcollections.

### CODE REVIEW

**Comment by Gankro** [comment]
Hmm, is this the sort of thing that needs to hang around forever, or is just pending a new snapshot?

**Comment by wesleywiser** [comment]
@Gankro I'm not sure. Sorry, I'm new to working on rust. What's the effect of taking a new snapshot?

**Comment by Gankro** [comment]
Stage0 is building the current source from a snapshot binary build of the source from a while back. This often means stage0 has to do some trippy stuff to bridge the gap between today's and yesterday's Rust. Taking a new snapshot makes the gap smaller and thus reduces the amount of conditional things the source needs to include to handle this. If this doesn't pop up in stage1 or 2 I assume that's the problem, but since 1.0 I haven't kept up on what the snapshot "process" is.

**Comment by Gankro** [comment]
Basically I like the idea of cleaning up the output, but not if it's just making more work for a temporary thing. We're doing triage tomorrow, and I'll look into what's up with this then.

Thanks for the contribution, though!

**Comment by wesleywiser** [comment]
Sure. Thanks for the explanation! I initially only noticed the warnings
during stage0 so I'm not sure if they also occur in later stages. If they
don't, I assume this is a non issue then?

**Comment by alexcrichton** [comment]
I think this isn't related to the stage0 build but rather unit tests (which means that these need to stick around for awhile), just to clarify! I'll add a few comments inline.

File: src/libcollections/lib.rs [hunk 1]
```diff
@@ -37,11 +37,10 @@
 #![feature(unsafe_no_drop_flag, filling_drop)]
 #![feature(step_by)]
 #![feature(str_char)]
-#![feature(str_words)]
+#![cfg_attr(not(test), feature(str_words))]
```
Comment (alexcrichton) [inline src/libcollections/lib.rs]: Could you move this down near the other `cfg_attr` features?

File: src/libcollections/slice.rs [hunk 2]
```diff
@@ -79,10 +79,15 @@
 #![doc(primitive = "slice")]
 #![stable(feature = "rust1", since = "1.0.0")]
 
+#[cfg(not(test))]
```
Comment (alexcrichton) [inline src/libcollections/slice.rs]: These attributes are unfortunately kinda sprinkled and make this tough to read, I think it's fine for this module to have `#![allow(unused_imports)]` with a comment explaining about all the `#[cfg(not(test))]` code

File: src/libcollections/str.rs [hunk 3]
```diff
@@ -51,13 +51,22 @@ use self::RecompositionState::*;
 use self::DecompositionType::*;
 
 use core::clone::Clone;
-use core::iter::{Iterator, Extend};
+use core::iter::Iterator;
+#[cfg(not(test))]
```
Comment (alexcrichton) [inline src/libcollections/str.rs]: Same comment as above here as well.

**Comment by wesleywiser** [comment]
@alexcrichton Thanks for the feedback! I've pushed a new commit with your suggestions.

File: src/libcollections/slice.rs [hunk 4]
```diff
@@ -79,6 +79,10 @@
 #![doc(primitive = "slice")]
 #![stable(feature = "rust1", since = "1.0.0")]
 
+//Many of the usings in this module are only used in the test configuration.
```
Comment (alexcrichton) [inline src/libcollections/slice.rs]: Conventionally we have a space after the last slash (e.g. `// foo`)

Comment (wesleywiser) [inline src/libcollections/slice.rs]: Ah, sorry

**Comment by wesleywiser** [comment]
@alexcrichton Fixed. Thanks!

**Pull request merged**